### PR TITLE
feat(cache): added more cache interactions 

### PR DESCRIPTION
**Pull request by jeabakker**

### CODE REVIEW

File: docs/design/events.rst [hunk 1]
```diff
@@ -211,6 +211,31 @@ Parameters:
 The function will return ``false`` if any of the selected handlers returned
 ``false`` and the event is stoppable, otherwise it will return ``true``.
 
+.. _design/events#event-sequence:
+
+Trigger an Elgg Event sequence
+------------------------------
+
+Instead of triggering the ``:before`` and ``:after`` event manually, its possible to trigger an event sequence. This will trigger
```
Comment (jdalsem) [inline docs/design/events.rst]: it's

File: docs/design/events.rst [hunk 2]
```diff
@@ -211,6 +211,31 @@ Parameters:
 The function will return ``false`` if any of the selected handlers returned
 ``false`` and the event is stoppable, otherwise it will return ``true``.
 
+.. _design/events#event-sequence:
+
+Trigger an Elgg Event sequence
+------------------------------
+
+Instead of triggering the ``:before`` and ``:after`` event manually, its possible to trigger an event sequence. This will trigger 
+the ``:before`` event, than the actual event and finaly the ``:after`` event.
```
Comment (jdalsem) [inline docs/design/events.rst]: then... finally

File: docs/design/events.rst [hunk 3]
```diff
@@ -211,6 +211,31 @@ Parameters:
...
+
+.. code-block:: php
+
+	elgg()->events->triggerSequence($event, $type, $object, $callable);
+
+When called with for exampe ``'cache:clear', 'system'`` the following three events are triggered
```
Comment (jdalsem) [inline docs/design/events.rst]: example

File: engine/classes/Elgg/Cache/CompositeCache.php [hunk 4]
```diff
@@ -308,6 +341,10 @@ protected function buildFileSystemDriver() {
 		if (!$path) {
 			return null;
 		}
+		
+		// make a sepatate folder for Stash caches
+		// because Stash assumes all files/folder as made by Stash
```
Comment (jdalsem) [inline engine/classes/Elgg/Cache/CompositeCache.php]: are*

Comment (jdalsem) [reply engine/classes/Elgg/Cache/CompositeCache.php]: folders*

File: engine/classes/Elgg/Cache/SimpleCache.php [hunk 5]
```diff
@@ -143,18 +143,65 @@ private function getPath() {
...
-	 * lastupdate time to 0 for every valid viewtype.
+	 * Deletes all cached views in the simplecache
+	 *
+	 * @param bool $include_clear (internal) Also clear cache
 	 *
 	 * @return bool
 	 */
-	public function invalidate() {
+	public function invalidate(bool $include_clear = null) {
+		$this->config->save('simplecache_lastupdate', time());
```
Comment (jdalsem) [inline engine/classes/Elgg/Cache/SimpleCache.php]: add @todo remove in 4.0

File: engine/classes/Elgg/Cache/SystemCache.php [hunk 6]
```diff
@@ -70,7 +70,7 @@ public function load($type) {
 			return;
 		}
 		
-		$cached_data = $this->cache->load($type);
+		$cached_data = $this->cache->load($type, null);
```
Comment (jdalsem) [inline engine/classes/Elgg/Cache/SystemCache.php]: Why was this needed?

File: engine/classes/Elgg/Config.php [hunk 7]
```diff
@@ -63,7 +63,7 @@
  * @property bool          $installer_running
  * @property string        $language                   				Site language code
  * @property string[]      $language_to_locale_mapping 				A language to locale mapping (eg. 'en' => ['en_US'] or 'nl' => ['nl_NL'])
- * @property int           $lastcache								The last cache timestamp
+ * @property int           $lastcache								The timestamp the cache was last invalidated/cleared
```
Comment (jdalsem) [inline engine/classes/Elgg/Config.php]: remove 'cleared'

File: engine/classes/Elgg/EventsService.php [hunk 8]
```diff
@@ -193,7 +196,6 @@ public function triggerSequence($event, $object_type, $object = null, callable $
 	 * @return bool
 	 *
 	 * @see EventsService::trigger()
-	 * @see elgg_trigger_deprecated_event()
```
Comment (jdalsem) [inline engine/classes/Elgg/EventsService.php]: huh? it still applies here i guess

File: engine/classes/Elgg/EventsService.php [hunk 9]
```diff
@@ -202,4 +204,27 @@ public function triggerDeprecated($event, $object_type, $object = null, $message
...
+	 * @param string   $object_type The object type
+	 * @param mixed    $object      The object involved in the event
+	 * @param callable $callable    Callable to run on successful event, before event:after
+	 * @param string   $message     The deprecation message
+	 * @param string   $version     Human-readable *release* version: 1.9, 1.10, ...
+	 *
+	 * @return bool
+	 *
+	 * @see EventsService::trigger()
+	 * @see elgg_trigger_deprecated_event()
```
Comment (jdalsem) [inline engine/classes/Elgg/EventsService.php]: probably should have removed this

File: engine/classes/Elgg/Cache/SimpleCache.php [hunk 10]
```diff
@@ -143,18 +143,65 @@ private function getPath() {
...
-	 * Deletes all cached views in the simplecache and sets the lastcache and
-	 * lastupdate time to 0 for every valid viewtype.
+	 * Deletes all cached views in the simplecache
+	 *
+	 * @param bool $include_clear (internal) Also clear cache
 	 *
 	 * @return bool
 	 */
-	public function invalidate() {
+	public function invalidate(bool $include_clear = null) {
```
Comment (jdalsem) [inline engine/classes/Elgg/Cache/SimpleCache.php]: add @todo to remove `include_clear` param in 4.0

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/Cache/SystemCache.php (hunk 6 above)
Comment (jeabakker) [reply engine/classes/Elgg/Cache/SystemCache.php]: left over, oops

File: engine/classes/Elgg/EventsService.php (hunk 8 above)
Comment (jeabakker) [reply engine/classes/Elgg/EventsService.php]: done

File: engine/classes/Elgg/EventsService.php (hunk 9 above)
Comment (jeabakker) [reply engine/classes/Elgg/EventsService.php]: indeed

**Pull request merged**